### PR TITLE
Scalar arithmetic should return error when overflows.

### LINKED ISSUE #5810: Arithmetic overflow in aggregate_statistics::AggregateStatistics when MAX-MIN overflows
### Describe the bug

When running statistics, selectivity is calculated by the distance between selected range and total range. See analyze_expr_scalar_comparison.

While calculating the distance, scalar::distance is simply subtracting MIN from MAX and it's possible it panicked or returns a negative number when it overflows.

### To Reproduce

Create a parquet file with max value i64::MAX and min value i64::MIN.
Run a query like:
```rust
    let ctx = SessionContext::new();
    let df = ctx.read_parquet("<file>.parquet", ParquetReadOptions::default()).await?;
    let df = df
        .filter(col("value").lt(lit(0 as i64)))?
        .aggregate(vec![], vec![max(col("value"))])?;
    df.show().await?;
```

### Expected behavior

It shows the result.

### Additional context

It panicked when running in debug mode.
```
thread 'main' panicked at 'attempt to subtract with overflow', ~/repo/arrow-datafusion/datafusion/common/src/scalar.rs:1816:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/8460ca823e8367a30dda430efda790588b8c84d3/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/8460ca823e8367a30dda430efda790588b8c84d3/library/core/src/panicking.rs:64:14
   2: core::panicking::panic
             at /rustc/8460ca823e8367a30dda430efda790588b8c84d3/library/core/src/panicking.rs:114:5
   3: datafusion_common::scalar::ScalarValue::sub
             at ~/repo/arrow-datafusion/datafusion/common/src/scalar.rs:1816:9
   4: datafusion_common::scalar::ScalarValue::distance
             at ~/repo/arrow-datafusion/datafusion/common/src/scalar.rs:1885:13
   5: datafusion_physical_expr::expressions::binary::analyze_expr_scalar_comparison
             at ~/repo/arrow-datafusion/datafusion/physical-expr/src/expressions/binary.rs:861:57
   6: <datafusion_physical_expr::expressions::binary::BinaryExpr as datafusion_physical_expr::physical_expr::PhysicalExpr>::analyze
             at ~/repo/arrow-datafusion/datafusion/physical-expr/src/expressions/binary.rs:732:25
   7: <datafusion::physical_plan::filter::FilterExec as datafusion::physical_plan::ExecutionPlan>::statistics
             at ~/repo/arrow-datafusion/datafusion/core/src/physical_plan/filter.rs:183:28
   8: datafusion::physical_optimizer::aggregate_statistics::take_optimizable
             at ~/repo/arrow-datafusion/datafusion/core/src/physical_optimizer/aggregate_statistics.rs:127:37
   9: <datafusion::physical_optimizer::aggregate_statistics::AggregateStatistics as datafusion::physical_optimizer::optimizer::PhysicalOptimizerRule>::optimize
             at ~/repo/arrow-datafusion/datafusion/core/src/physical_optimizer/aggregate_statistics.rs:56:41
  10: datafusion::physical_plan::planner::DefaultPhysicalPlanner::optimize_internal
             at ~/repo/arrow-datafusion/datafusion/core/src/physical_plan/planner.rs:1794:24
  11: <datafusion::physical_plan::planner::DefaultPhysicalPlanner as datafusion::physical_plan::planner::PhysicalPlanner>::create_physical_plan::{{closure}}
             at ~/repo/arrow-datafusion/datafusion/core/src/physical_plan/planner.rs:427:17
  12: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/8460ca823e8367a30dda430efda790588b8c84d3/library/core/src/future/future.rs:125:9
  13: <datafusion::execution::context::DefaultQueryPlanner as datafusion::execution::context::QueryPlanner>::create_physical_plan::{{closure}}
             at ~/repo/arrow-datafusion/datafusion/core/src/execution/context.rs:1175:13
  14: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/8460ca823e8367a30dda430efda790588b8c84d3/library/core/src/future/future.rs:125:9
  15: datafusion::execution::context::SessionState::create_physical_plan::{{closure}}
             at ~/repo/arrow-datafusion/datafusion/core/src/execution/context.rs:1670:13
  16: datafusion::dataframe::DataFrame::create_physical_plan::{{closure}}
             at ~/repo/arrow-datafusion/datafusion/core/src/dataframe.rs:99:60
  17: datafusion::dataframe::DataFrame::collect::{{closure}}
             at ~/repo/arrow-datafusion/datafusion/core/src/dataframe.rs:663:47
  18: datafusion::dataframe::DataFrame::show::{{closure}}
             at ~/repo/arrow-datafusion/datafusion/core/src/dataframe.rs:681:37
  19: rust_sample::main::{{closure}}
             at ./src/main.rs:33:14
  20: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/park.rs:283:63
  21: tokio::runtime::coop::with_budget
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/coop.rs:107:5
  22: tokio::runtime::coop::budget
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/coop.rs:73:5
  23: tokio::runtime::park::CachedParkThread::block_on
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/park.rs:283:31
  24: tokio::runtime::context::BlockingRegionGuard::block_on
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/context.rs:315:13
  25: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/scheduler/multi_thread/mod.rs:66:9
  26: tokio::runtime::runtime::Runtime::block_on
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.26.0/src/runtime/runtime.rs:304:45
  27: rust_sample::main
             at ./src/main.rs:69:5
  28: core::ops::function::FnOnce::call_once
             at /rustc/8460ca823e8367a30dda430efda790588b8c84d3/library/core/src/ops/function.rs:250:5
```

While it doesn't panic in release mode but it's subject to pick sub-optimal plans because of the wrong stats.

### PR DESCRIPTION
**Pull request by zhzy0077**
# Which issue does this PR close?

Closes #5810.

# Rationale for this change

Repro in the bug itself. ScalarValue returns an error for many arithmetic errors, but not overflows. 

# What changes are included in this PR?

Use `checked_*` and check results.

# Are these changes tested?

Tests added.

# Are there any user-facing changes?

No.

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Thank you @zhzy0077  -- this looks like a great improvement to me

Can you also please add a test that ensures the overflow behavior is consistent with the arrow compute kernels (I can't remember if they error on overflow). Perhaps following the model of:

https://github.com/apache/arrow-datafusion/blob/ab2ad35e821e754c5ba20e92da0f6cacf205b831/datafusion/common/src/scalar.rs#L4896-L4913

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -573,12 +573,43 @@ macro_rules! primitive_op {
...
+                        "Overflow while calculating ScalarValue.".to_string(),
+                    ))
+                }
+            }
         }
     };
 }
 
+macro_rules! primitive_checked_op {
+    ($LEFT:expr, $RIGHT:expr, $OPERATION:tt, Float64) => {
```
Comment (comphead) [inline datafusion/common/src/scalar.rs:593]: cool!

**Comment by zhzy0077** [comment]
> Thank you @zhzy0077 -- this looks like a great improvement to me
> 
> Can you also please add a test that ensures the overflow behavior is consistent with the arrow compute kernels (I can't remember if they error on overflow). Perhaps following the model of:
> 
> https://github.com/apache/arrow-datafusion/blob/ab2ad35e821e754c5ba20e92da0f6cacf205b831/datafusion/common/src/scalar.rs#L4896-L4913

Thank you. Added a few tests in `scalar_add_overflow_test`. Not sure if this covers all your concerns.

**Comment by tustvold** [comment]
> mostly for consistency with the semantics of arrow-rs

So this is not consistent with the arrow kernels that DataFusion makes use of, in particular, DataFusion is currently using the unchecked arithmetic kernels, which do not return an error on overflow. So this PR will introduce inconsistency between ScalarValue arithmetic, and any arithmetic involving arrays.

The major reason I'm a little apprehensive about changing this is that the checked kernels are at least an order of magnitude slower in the presence of nulls, as LLVM can't vectorise them correctly

```
add(0.1)                time:   [7.4481 µs 7.4539 µs 7.4608 µs]
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

add_checked(0.1)        time:   [101.62 µs 106.31 µs 111.40 µs]
```

I definitely think whatever semantics we settle on should be consistent and well documented, but I don't have a strong opinion what it should be. However, I do feel that changing the current semantics warrants some communication due to the major performance regression it would entail.

**Comment by zhzy0077** [comment]
Thank you for the detailed explanation @tustvold. What do you think if we introduce the other set of operands, say "add_checked", "mul_checked", etc.?

**Comment by zhzy0077** [comment]
@alamb @tustvold PR is updated. `add/sub` methods are no longer touched and `add_checked/sub_checked` are added. Could you help to take another look?

**Comment by waynexia** [comment]
Test error:

> ---- scalar::tests::scalar_sub_trait_int64_overflow_test stdout ----
thread 'scalar::tests::scalar_sub_trait_int64_overflow_test' panicked at 'attempt to subtract with overflow', datafusion/common/src/scalar.rs:1934:9

It looks like the methods called in unit tests should also be `*_checked`

**Review by waynexia — APPROVED** [review]
Looks good, thanks

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -3664,6 +3746,100 @@ mod tests {
...
+        check_scalar_add_overflow::<UInt64Type>(
+            ScalarValue::UInt64(Some(u64::MAX)),
+            ScalarValue::UInt64(Some(u64::MAX)),
+        );
+
+        Ok(())
+    }
+
+    // Verifies that ScalarValue has the same behavior with compute kernal when it overflows.
+    fn check_scalar_add_overflow<T>(left: ScalarValue, right: ScalarValue)
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:3828]: Thank you for this test

**Review by alamb — COMMENTED** [review]
Thank you @zhzy0077  and @waynexia for the review

**Pull request merged**